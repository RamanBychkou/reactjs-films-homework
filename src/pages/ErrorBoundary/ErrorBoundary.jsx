import React, { Component } from 'react';
import PropTypes from 'prop-types';
import MovieSelectors from '../../components/MovieSelectors/index';
import Header from '../../components/Header/index';
import style from './ErrorBoundary.scss';

class ErrorBoundary extends Component {
  constructor(props) {
    super(props);
    this.state = { hasError: '' };
  }

  static getDerivedStateFromProps(nextProps) {
    if (nextProps.error !== '') {
      return {
        hasError: true,
        typeError: nextProps.error,
      };
    }
    return {
      hasError: false,
      typeError: null,
    };
  }

  shouldComponentUpdate(prevProps) {
    const { error } = this.props;
    if (prevProps.error !== error) {
      return true;
    }
    return false;
  }

  render() {
    const { hasError, typeError } = this.state;
    const {
      children, genres, error, clearError, fetchListMovies, fetchSearchResults,
    } = this.props;
    if (hasError === true) {
      if (typeError !== 'Something Wrong') {
        return (
          <div>
            <Header fetchSearchResults={fetchSearchResults} error={error} clearError={clearError} />
            <div className={style.container}>
              <div className="title">
                <h1>Nothing found</h1>
              </div>
              <MovieSelectors
                genres={genres}
                error={error}
                clearError={clearError}
                fetchListMovies={fetchListMovies}
              />
              <img
                src="https://cdn.shopify.com/s/files/1/1164/8172/products/funny-insomnia-mug-error-404-am-sleep-not-found-11oz-black-coffee-mugs_175_1024x1024.jpg?v=1537210909"
                alt="404 Not Found"
              />
            </div>
          </div>
        );
      }
      return <h1>Something Wrong</h1>;
    }
    return children;
  }
}
ErrorBoundary.defaultProps = {
  error: '',
  genres: [],
  children: [],
};

ErrorBoundary.propTypes = {
  error: PropTypes.string,
  genres: PropTypes.arrayOf(PropTypes.object),
  fetchListMovies: PropTypes.func.isRequired,
  clearError: PropTypes.func.isRequired,
  children: PropTypes.arrayOf(PropTypes.any),
  fetchSearchResults: PropTypes.func.isRequired,
};
export default ErrorBoundary;
