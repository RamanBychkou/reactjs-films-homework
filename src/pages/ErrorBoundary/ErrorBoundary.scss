.container {
  max-width: 82rem;
  margin: 0 auto;
  padding-top: 5rem;
  text-align: center;
  display: flex;
  flex-flow: column nowrap;
  height: 100%;

  .imgWrapper {
    min-height: 92%;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  img {
    width: 45%;
  }
}
