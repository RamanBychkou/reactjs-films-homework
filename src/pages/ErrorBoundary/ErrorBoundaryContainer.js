import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import ErrorBoundary from './ErrorBoundary';
import itemsReducer from '../../modules/Error/errorAction';
import requestsFilms from '../../utils/requests';

const mapStateToProps = state => state;
export const mapStateToDispatch = dispatch => ({
  clearError: boolean => dispatch(itemsReducer(boolean)),
  getMainMovieDetails: id => dispatch(requestsFilms.getMainMovieDetails(id)),
  fetchSearchResults: query => dispatch(requestsFilms.fetchSearchResults(query)),
  fetchListMovies: query => dispatch(requestsFilms.fetchListMovies(query)),
});
export default withRouter(
  connect(
    mapStateToProps,
    mapStateToDispatch,
  )(ErrorBoundary),
);
