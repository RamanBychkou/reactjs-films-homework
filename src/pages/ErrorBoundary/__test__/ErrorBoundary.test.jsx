import React from 'react';
import { MemoryRouter as Router } from 'react-router-dom';
import TestRender from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import ErrorBoundary from '../index';
import { mapStateToDispatch } from '../ErrorBoundaryContainer';

describe('getDerivedStateFromProps', () => {
  it('getDerivedStateFromProps: has error', async () => {
    const error = 'Something Wrong';
    const initial = {
      error,
    };
    const action = await ErrorBoundary.WrappedComponent.getDerivedStateFromProps.call(null, {
      ...initial,
    });

    const expected = {
      hasError: true,
      typeError: error,
    };
    expect(action).toEqual(expected);
  });

  it('getDerivedStateFromProps: error Nothing Found', async () => {
    const error = 'Something Wrong';
    const initial = {
      error,
    };
    const action = await ErrorBoundary.WrappedComponent.getDerivedStateFromProps.call(null, {
      ...initial,
    });

    const expected = {
      hasError: true,
      typeError: error,
    };
    expect(action).toEqual(expected);
  });

  it('getDerivedStateFromProps:  error = ""', async () => {
    const error = '';
    const initial = {
      error,
    };
    const action = await ErrorBoundary.WrappedComponent.getDerivedStateFromProps.call(null, {
      ...initial,
    });

    const expected = {
      hasError: false,
      typeError: null,
    };
    expect(action).toEqual(expected);
  });
});

describe('render', () => {
  it('render correctly: Something Wrong error', () => {
    const error = 'Something Wrong';
    const initial = {
      error,
    };
    let store = configureStore();
    store = store(initial);
    const render = TestRender.create(
      <Provider store={store}>
        <Router>
          <ErrorBoundary />
        </Router>
      </Provider>,
    );
    expect(render).toMatchSnapshot();
  });

  it('render correctly: Nothing Found error', () => {
    const error = 'Nothing Found';
    const initial = {
      error,
    };
    let store = configureStore();
    store = store(initial);
    const render = TestRender.create(
      <Provider store={store}>
        <Router>
          <ErrorBoundary />
        </Router>
      </Provider>,
    );
    expect(render).toMatchSnapshot();
  });

  it('render correctly: error = ""', () => {
    const error = '';
    const initial = {
      error,
    };
    let store = configureStore();
    store = store(initial);
    const render = TestRender.create(
      <Provider store={store}>
        <Router>
          <ErrorBoundary />
        </Router>
      </Provider>,
    );
    expect(render).toMatchSnapshot();
  });
});

describe('ComponentDidUpdate', () => {
  it('ComponentDidUpdate: error = ""', async () => {
    const error = '';
    const initial = {
      props: { error },
    };
    const action = await ErrorBoundary.WrappedComponent.prototype.shouldComponentUpdate.call(
      initial,
      {
        error,
      },
    );

    expect(action).toBeFalsy();
  });

  it('ComponentDidUpdate: error = true', async () => {
    const error = 'true';
    const initial = {
      props: { error },
    };
    const action = await ErrorBoundary.WrappedComponent.prototype.shouldComponentUpdate.call(
      initial,
      {
        error: '',
      },
    );

    expect(action).toBeTruthy();
  });
});

describe('test MapDispatchToProps', () => {
  const state = {
    fetchListMovies: id => id,
    getMainMovieDetails: id => id,
    fetchSearchResults: query => query,
    clearError: query => query,
  };
  it('test all descriptors', () => {
    const keys = Object.keys(state);
    const id = 'test';
    keys.forEach((curr) => {
      const dispatch = jest.fn(() => state[curr]);
      const result = mapStateToDispatch(dispatch)[curr]();
      expect(result(id)).toEqual(id);
    });
  });
});
