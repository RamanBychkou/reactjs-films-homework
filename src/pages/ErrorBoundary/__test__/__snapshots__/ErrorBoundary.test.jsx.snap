// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render render correctly: Nothing Found error 1`] = `
<div>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <h1
        className="h1"
      >
        FILMS
      </h1>
      <form
        action=""
        onSubmit={[Function]}
      >
        <input
          className="searchField"
          tabIndex="0"
          type="search"
        />
        <button
          className="btnSearch"
          type="submit"
        >
          ⚲
        </button>
      </form>
    </div>
  </header>
  <div
    className="container"
  >
    <div
      className="title"
    >
      <h1>
        Nothing found
      </h1>
    </div>
    <div
      className="moviesSelector"
    >
      <nav>
        <ul>
          <li>
            <a
              href="/trending"
              onClick={[Function]}
            >
              <button
                href="#"
                onClick={[Function]}
                type="button"
              >
                Trending
              </button>
            </a>
          </li>
          <li>
            <a
              href="/top_rated"
              onClick={[Function]}
            >
              <button
                href="#"
                onClick={[Function]}
                type="button"
              >
                Top Rated
              </button>
            </a>
          </li>
          <li>
            <a
              href="/coming_soon"
              onClick={[Function]}
            >
              <button
                href="#"
                onClick={[Function]}
                type="button"
              >
                Coming Soon
              </button>
            </a>
          </li>
          <li>
            <select
              id=""
              name="genre"
              onChange={[Function]}
            >
              <option
                value=""
              >
                Genre
              </option>
            </select>
          </li>
        </ul>
      </nav>
      <div
        className="typeView"
      >
        <button
          id="cards"
          onClick={[Function]}
          type="button"
        >
          ⚏
        </button>
        <button
          id="lines"
          onClick={[Function]}
          type="button"
        >
          ⚌
        </button>
      </div>
    </div>
    <img
      alt="404 Not Found"
      src="https://cdn.shopify.com/s/files/1/1164/8172/products/funny-insomnia-mug-error-404-am-sleep-not-found-11oz-black-coffee-mugs_175_1024x1024.jpg?v=1537210909"
    />
  </div>
</div>
`;

exports[`render render correctly: Something Wrong error 1`] = `
<h1>
  Something Wrong
</h1>
`;

exports[`render render correctly: error = "" 1`] = `null`;
